### ISSUE TITLE
Max supply query doesn't work for addresses with large tx volume

### ISSUE THREAD

**Issue by lovetillion** [Panmoni/tokenstork#14]
there is currently a bug with the gql query causing market cap and max supply to be incorrect (at least for XLV) this value is absolutely fixed at 79000.57996200 from the burns done on June 5th and should be 61000.57996200 after the burns done today

The issue is with the qrstt2d9djm5pnh65jrfw3kwfwesnnafnyvqamdp2g address which has generated Main (31,700) transactions and CashTokens (FT)  (15,443) transactions on cauldron micro LPs which is likely exhausting the limits of the query responses

![image](https://github.com/user-attachments/assets/c7a99f66-d13e-4d1a-853f-72d5b6a1cf9c)

**Comment by georgedonnelly** [comment]
Hmm ok, thanks. I can't guarantee a quick analysis of this but it's on my radar.

**Comment by lovetillion** [comment]
All good my main flag for you is that the logic and query are fine the issue is that the query response can't handle the amount of tx volume on that address. Also just for curiosity sake even the blockchain nodes themselves reject transactions trying to move too many cashtokens at the same time on that address even if I tried to pay ~$1 in fees the byte size gets too big and is rejected so the broadcast doesn't work. I hope that participating in this community we were able to give more than we took and identified some areas for improvement but are moving to other blockchains now so also won't be able to put any focus into this one.

**Comment by georgedonnelly** [comment]
If that's the case then it may be an issue with the chaingraph server, which I am borrowing from mainnet_pat and which is quite expensive and resource-intensive to host. Yes, indeed, we are kind of in the same boat. Did you report the issue with the transaction size to cashtoken_devs by chance?

**Comment by lovetillion** [comment]
I agree definitely not a trivial problem and I'm sure those constraints are there with good reason it just invalidates the original solution which we knew was time limited. I did flag it in one of the dev channels and haven't heard back from anyone there or on twitter so we can just leave this open for now. Hope your other ventures are doing well and it has been great working with you.

**Comment by georgedonnelly** [comment]
Worth an update here. The stack that caused this bug doesn't exist anymore — TokenStork is beingrewritten around a self-hosted indexer (archival BCHN + BlockBook + Postgres + Rust sync workers, all on one VPS). No Chaingraph in the pipeline at all, so the GQL response-size ceiling on high-volume addresses like qrstt2d9djm5pnh65jrfw3kwfwesnnafnyvqamdp2g is no longer in the picture.

In the new design, circulating supply is computed by summing unspent token UTXOs per category directly off the indexed UTXO set — no per-address enumeration involved, so transaction volume at any single address is irrelevant to the math.

That said, I'm not closing this yet. The BlockBook half of the stack (Phase 2d in our internal plan) isn't deployed on the production VPS yet, so current_supply is null in the UI today. Once BlockBook is live I'll re-run enrichment against XLV specifically and confirm the value matches the known post-burn total (~61000.57996200 as of your Sep 9 2024 burn, modulo anything burned since). Will report back here with the number and close on green. If it still disagrees after that, the bug will have moved into our own enrichment logic and I'll dig in there.

Thanks again for the detailed report — and for the heads-up about the byte-size limit on big multi-token spends; that one's definitely a protocol-layer conversation, not something we can fix from the indexer side.

**Comment by georgedonnelly** [comment]
Fixed. tokenstork's enrichment pipeline no longer goes through Chaingraph at all — the architecture moved to a self-hosted BCHN + BlockBook stack on a single VPS, and the supply walker now reads the indexed UTXO set directly via BlockBook's `/api/v2/address/<category>?details=txs` endpoint with paginated tx walking. There's no per-address enumeration anywhere in the path, so the 31,700-tx response-size ceiling that originally clipped XLV's supply read is structurally absent.

Verified live (2026-05-02) on the production database for category `3270f6641a815709c9c03e75e189563c1bb2f0814d8644a2181da896fb3405d0`:

- `current_supply` = `155108537294` raw → **1551.08537294 XLV** (decimals=8)
- `holder_count` = 98
- `live_utxo_count` = 138
- `verified_at` = 2026-05-02 19:57 UTC

The original report quoted **61000.57996200 XLV after the 2024-09-09 burns**. 19 months on, supply is 1551 — consistent with continued burn activity over that window, and a real, finite, calculable number rather than a truncated read. The qrstt2d9… address that originally tripped the response-size limit doesn't even appear in our `token_holders` table — but the supply walker doesn't depend on enumerating it, because BlockBook's UTXO index is queried directly.

Closing.
